File: Source/Game/Items/ItemSpatialGrid.cs
using System.Collections.Generic;
using WCSharp.Api;

public static class ItemSpatialGrid
{
    private const float CELL_SIZE = 128f;
    private const int GRID_WIDTH = 64;
    private const int GRID_HEIGHT = 64;
    private const int X_OFFSET = 32;
    private const int Y_OFFSET = 32;
    private const int INVALID_CELL = -1;

    private static readonly List<Kibble>[] kibbleCells = new List<Kibble>[GRID_WIDTH * GRID_HEIGHT];
    private static readonly List<item>[] itemCells = new List<item>[GRID_WIDTH * GRID_HEIGHT];

    private static int GetCellIndex(float x, float y)
    {
        int cellX = (int)(x / CELL_SIZE) + X_OFFSET;
        int cellY = (int)(y / CELL_SIZE) + Y_OFFSET;
        if (cellX < 0 || cellX >= GRID_WIDTH || cellY < 0 || cellY >= GRID_HEIGHT)
            return INVALID_CELL;
        return cellY * GRID_WIDTH + cellX;
    }

    public static void RegisterKibble(Kibble kibble)
    {
        int index = GetCellIndex(kibble.Item.X, kibble.Item.Y);
        if (index == INVALID_CELL) return;
        if (kibbleCells[index] == null)
            kibbleCells[index] = new List<Kibble>();
        kibbleCells[index].Add(kibble);
    }

    public static void UnregisterKibble(Kibble kibble)
    {
        int index = GetCellIndex(kibble.Item.X, kibble.Item.Y);
        if (index == INVALID_CELL) return;
        kibbleCells[index]?.Remove(kibble);
    }

    public static void RegisterItem(item item)
    {
        int index = GetCellIndex(item.X, item.Y);
        if (index == INVALID_CELL) return;
        if (itemCells[index] == null)
            itemCells[index] = new List<item>();
        itemCells[index].Add(item);
    }

    public static void UnregisterItem(item item)
    {
        int index = GetCellIndex(item.X, item.Y);
        if (index == INVALID_CELL) return;
        itemCells[index]?.Remove(item);
    }

    public static List<Kibble> GetNearbyKibbles(float x, float y)
    {
        int index = GetCellIndex(x, y);
        if (index == INVALID_CELL) return null;
        return kibbleCells[index];
    }

    public static List<item> GetNearbyItems(float x, float y)
    {
        int index = GetCellIndex(x, y);
        if (index == INVALID_CELL) return null;
        return itemCells[index];
    }

    public static void KittyItemPickup(Kitty kitty)
    {
        var kibbleList = GetNearbyKibbles(kitty.Unit.X, kitty.Unit.Y);
        var itemList = GetNearbyItems(kitty.Unit.X, kitty.Unit.Y);

        if (kibbleList != null && kibbleList.Count > 0)
        {
            for (int i = 0; i < kibbleList.Count; i++)
            {
                var k = kibbleList[i];
                if (k == null) continue;
                kitty.Unit.AddItem(k.Item);
            }
        }

        if (itemList != null && itemList.Count > 0)
        {
            for (int i = 0; i < itemList.Count; i++)
            {
                var item = itemList[i];
                if (item == null) continue;
                if (item.IsOwned) continue;
                kitty.Unit.AddItem(item);
                break;
            }
        }
    }
}

File: src/Game/Items/ItemSpatialGrid.ts
import { Item } from 'w3ts'
import { Kitty } from '../Entities/Kitty/Kitty'
import { Kibble } from './Kibble'

export class ItemSpatialGrid {
    private static CELL_SIZE = 128
    private static kibbleCells: Map<string, Kibble[]> = new Map()
    private static itemCells: Map<string, Item[]> = new Map()

    private static getCellKey(x: number, y: number): string {
        const cellX = Math.floor(x / ItemSpatialGrid.CELL_SIZE)
        const cellY = Math.floor(y / ItemSpatialGrid.CELL_SIZE)
        return `${cellX},${cellY}`
    }

    public static RegisterKibble = (kibble: Kibble) => {
        if (!kibble.Item) return // If the item is not valid... no reason to register i guess
        const cellKey = ItemSpatialGrid.getCellKey(kibble.Item.x, kibble.Item.y)
        let list = ItemSpatialGrid.kibbleCells.get(cellKey)
        if (!list) ItemSpatialGrid.kibbleCells.set(cellKey, (list = []))
        list.push(kibble)
    }

    public static UnregisterKibble = (kibble: Kibble) => {
        if (!kibble.Item) return // If the item is not valid... no reason to unregister i guess
        const cellKey = ItemSpatialGrid.getCellKey(kibble.Item.x, kibble.Item.y)
        const list = ItemSpatialGrid.kibbleCells.get(cellKey) || []

        ItemSpatialGrid.kibbleCells.set(
            cellKey,
            list.filter(i => i !== kibble)
        )
    }

    public static RegisterItem = (item: Item) => {
        const cellKey = ItemSpatialGrid.getCellKey(item.x, item.y)
        let list: Item[]
        if (!(list = ItemSpatialGrid.itemCells.get(cellKey)!)) ItemSpatialGrid.itemCells.set(cellKey, (list = []))
        list.push(item)
    }

    public static UnregisterItem = (item: Item) => {
        if (!item) return // If the item is not valid... no reason to unregister i guess
        const cellKey = ItemSpatialGrid.getCellKey(item.x, item.y)
        const list = ItemSpatialGrid.itemCells.get(cellKey) || []

        ItemSpatialGrid.itemCells.set(
            cellKey,
            list.filter(i => i !== item)
        )
    }

    public static GetNearbyKibbles = (x: number, y: number) => {
        const cellKey = ItemSpatialGrid.getCellKey(x, y)
        const list = ItemSpatialGrid.kibbleCells.get(cellKey)
        if (list) return list
        return null
    }

    public static GetNearbyItems = (x: number, y: number) => {
        const cellKey = ItemSpatialGrid.getCellKey(x, y)
        const list = ItemSpatialGrid.itemCells.get(cellKey)
        if (list) return list
        return null
    }

    public static KittyItemPickup = (kitty: Kitty) => {
        const kibbleList = ItemSpatialGrid.GetNearbyKibbles(kitty.Unit.x, kitty.Unit.y)
        const itemList = ItemSpatialGrid.GetNearbyItems(kitty.Unit.x, kitty.Unit.y)

        if (kibbleList && kibbleList.length > 0) {
            for (let i = 0; i < kibbleList.length; i++) {
                const k = kibbleList[i]
                if (!k?.Item) continue
                kitty.Unit.addItem(k.Item)
            }
        }

        if (itemList && itemList.length > 0) {
            for (let i = 0; i < itemList.length; i++) {
                const item = itemList[i]
                if (!item) continue
                if (item.isOwned()) continue
                kitty.Unit.addItem(item)
                break
            }
        }
    }
}
